### ISSUE TITLE
Leaf template caching with non-unique keys

### ISSUE THREAD

**Issue by foscomputerservices** [vapor/template-kit#25]
I'm struggling with the wrong leaf template being rendered in production (not in development). I'm almost 100% sure that this is a caching problem as if I turn caching off all works as expected.

In digging into TemplateRenderer, I find the following:

```swift
public func render(template: Data, _ context: TemplateData, file: String = "template") -> Future<View> {
    return Future.flatMap(on: container) {
        let hash = template.hashValue
        let ast: [TemplateSyntax]
        if let cached = self.astCache?.storage[hash] {
            ast = cached
        } else {
            let scanner = TemplateByteScanner(data: template, file: file)
            ast = try self.parser.parse(scanner: scanner)
            self.astCache?.storage[hash] = ast
        }
        ...
```

Can hashValue on Data really be used as a unique key? I seem to be getting a collision in two of my templates. In the attached debugger autos windows that is stopped at the "if let cached = ..." line above, you'll see that the hash is the same value for both template files, however, the template files are drastically different.

![screen shot 2018-06-21 at 12 12 47 pm](https://user-images.githubusercontent.com/2438421/41779746-9de9b008-75e7-11e8-84f7-05e5a87452ac.png)
![screen shot 2018-06-21 at 12 13 37 pm](https://user-images.githubusercontent.com/2438421/41779747-9e278982-75e7-11e8-952f-aa770ef3e21a.png)

(Please note: I migrated this issue from [Leaf #114](https://github.com/vapor/leaf/issues/114) as the issue really is with template-kit and not Leaf -- I apologize for incorrectly locating the issue originally)

**Comment by joeblau** [comment]
TemplateData needs to conform to the `Hashable` protocol and the corresponding function needs to be implemented.

**Comment by tanner0101** [comment]
This has been fixed in #31. The file path is now used. Templates that don't have a path are no longer cached. Thanks!
